// @genkit-ai/ai/lib/generate/action.mjs
import {
  __async,
  __spreadProps,
  __spreadValues
} from "../chunk-7LUMLJGI.mjs";
import {
  getStreamingCallback,
  runWithStreamingCallback,
  z
} from "@genkit-ai/core";
import { toJsonSchema } from "@genkit-ai/core/schema";
import { runInNewSpan, SPAN_TYPE_ATTR } from "@genkit-ai/core/tracing";
import * as clc from "colorette";
import { DocumentDataSchema } from "../document.js";
import {
  GenerateResponse,
  GenerateResponseChunk,
  tagAsPreamble
} from "../generate.js";
import {
  MessageSchema,
  ToolDefinitionSchema
} from "../model.js";
import { lookupToolByName, toToolDefinition } from "../tool.js";
const GenerateUtilParamSchema = z.object({
  /** A model name (e.g. `vertexai/gemini-1.0-pro`). */
  model: z.string(),
  /** Retrieved documents to be used as context for this generation. */
  docs: z.array(DocumentDataSchema).optional(),
  /** Conversation history for multi-turn prompting when supported by the underlying model. */
  messages: z.array(MessageSchema),
  /** List of registered tool names for this generation if supported by the underlying model. */
  tools: z.array(z.union([z.string(), ToolDefinitionSchema])).optional(),
  /** Configuration for the generation request. */
  config: z.any().optional(),
  /** Configuration for the desired output of the request. Defaults to the model's default output if unspecified. */
  output: z.object({
    format: z.union([z.literal("text"), z.literal("json"), z.literal("media")]).optional(),
    jsonSchema: z.any().optional()
  }).optional(),
  /** When true, return tool calls for manual processing instead of automatically resolving them. */
  returnToolRequests: z.boolean().optional()
});
function generateHelper(registry, input, middleware) {
  return __async(this, null, function* () {
    return yield runInNewSpan(
      {
        metadata: {
          name: "generate"
        },
        labels: {
          [SPAN_TYPE_ATTR]: "helper"
        }
      },
      (metadata) => __async(this, null, function* () {
        metadata.name = "generate";
        metadata.input = input;
        const output = yield generate(registry, input, middleware);
        metadata.output = JSON.stringify(output);
        return output;
      })
    );
  });
}
function generate(registry, rawRequest, middleware) {
  return __async(this, null, function* () {
    var _a, _b, _c, _d, _e, _f, _g;
    const model = yield registry.lookupAction(
      `/model/${rawRequest.model}`
    );
    if (!model) {
      throw new Error(`Model ${rawRequest.model} not found`);
    }
    if (((_a = model.__action.metadata) == null ? void 0 : _a.model.stage) === "deprecated") {
      console.warn(
        `${clc.bold(clc.yellow("Warning:"))} Model '${model.__action.name}' is deprecated and may be removed in a future release.`
      );
    }
    let tools;
    if ((_b = rawRequest.tools) == null ? void 0 : _b.length) {
      if (!((_d = (_c = model.__action.metadata) == null ? void 0 : _c.model.supports) == null ? void 0 : _d.tools)) {
        throw new Error(
          `Model ${rawRequest.model} does not support tools, but some tools were supplied to generate(). Please call generate() without tools if you would like to use this model.`
        );
      }
      tools = yield Promise.all(
        rawRequest.tools.map((toolRef) => __async(this, null, function* () {
          if (typeof toolRef === "string") {
            return lookupToolByName(registry, toolRef);
          } else if (toolRef.name) {
            return lookupToolByName(registry, toolRef.name);
          }
          throw `Unable to resolve tool ${JSON.stringify(toolRef)}`;
        }))
      );
    }
    const request = yield actionToGenerateRequest(rawRequest, tools);
    const accumulatedChunks = [];
    const streamingCallback = getStreamingCallback();
    const response = yield runWithStreamingCallback(
      streamingCallback ? (chunk) => {
        if (streamingCallback) {
          streamingCallback(
            new GenerateResponseChunk(chunk, {
              previousChunks: accumulatedChunks
            })
          );
        }
        accumulatedChunks.push(chunk);
      } : void 0,
      () => __async(this, null, function* () {
        const dispatch = (index, req) => __async(this, null, function* () {
          if (!middleware || index === middleware.length) {
            return yield model(req);
          }
          const currentMiddleware = middleware[index];
          return currentMiddleware(
            req,
            (modifiedReq) => __async(this, null, function* () {
              return dispatch(index + 1, modifiedReq || req);
            })
          );
        });
        return new GenerateResponse(yield dispatch(0, request), request);
      })
    );
    response.assertValid(request);
    const message = response.message;
    const toolCalls = message.content.filter((part) => !!part.toolRequest);
    if (rawRequest.returnToolRequests || toolCalls.length === 0) {
      return response.toJSON();
    }
    const toolResponses = [];
    let messages = [...request.messages, message];
    let newTools = rawRequest.tools;
    for (const part of toolCalls) {
      if (!part.toolRequest) {
        throw Error(
          "Tool request expected but not provided in tool request part"
        );
      }
      const tool = tools == null ? void 0 : tools.find(
        (tool2) => {
          var _a2;
          return tool2.__action.name === ((_a2 = part.toolRequest) == null ? void 0 : _a2.name);
        }
      );
      if (!tool) {
        throw Error(`Tool ${(_e = part.toolRequest) == null ? void 0 : _e.name} not found`);
      }
      if (tool.__action.metadata.type === "prompt") {
        const newPreamble = yield tool((_f = part.toolRequest) == null ? void 0 : _f.input);
        toolResponses.push({
          toolResponse: {
            name: part.toolRequest.name,
            ref: part.toolRequest.ref,
            output: `transferred to ${part.toolRequest.name}`
          }
        });
        messages = [
          ...tagAsPreamble(newPreamble.messages),
          ...messages.filter((m) => {
            var _a2;
            return !((_a2 = m == null ? void 0 : m.metadata) == null ? void 0 : _a2.preamble);
          })
        ];
        newTools = newPreamble.tools;
      } else {
        toolResponses.push({
          toolResponse: {
            name: part.toolRequest.name,
            ref: part.toolRequest.ref,
            output: yield tool((_g = part.toolRequest) == null ? void 0 : _g.input)
          }
        });
      }
    }
    const nextRequest = __spreadProps(__spreadValues({}, rawRequest), {
      messages: [
        ...messages,
        {
          role: "tool",
          content: toolResponses
        }
      ],
      tools: newTools
    });
    return yield generateHelper(registry, nextRequest, middleware);
  });
}
function actionToGenerateRequest(options, resolvedTools) {
  return __async(this, null, function* () {
    var _a, _b, _c;
    const out = {
      messages: options.messages,
      config: options.config,
      docs: options.docs,
      tools: (resolvedTools == null ? void 0 : resolvedTools.map((tool) => toToolDefinition(tool))) || [],
      output: {
        format: ((_a = options.output) == null ? void 0 : _a.format) || (((_b = options.output) == null ? void 0 : _b.jsonSchema) ? "json" : "text"),
        schema: toJsonSchema({
          jsonSchema: (_c = options.output) == null ? void 0 : _c.jsonSchema
        })
      }
    };
    if (!out.output.schema) delete out.output.schema;
    return out;
  });
}
function inferRoleFromParts(parts) {
  const uniqueRoles = /* @__PURE__ */ new Set();
  for (const part of parts) {
    const role = getRoleFromPart(part);
    uniqueRoles.add(role);
    if (uniqueRoles.size > 1) {
      throw new Error("Contents contain mixed roles");
    }
  }
  return Array.from(uniqueRoles)[0];
}
function getRoleFromPart(part) {
  if (part.toolRequest !== void 0) return "model";
  if (part.toolResponse !== void 0) return "tool";
  if (part.text !== void 0) return "user";
  if (part.media !== void 0) return "user";
  if (part.data !== void 0) return "user";
  throw new Error("No recognized fields in content");
}
export {
  GenerateUtilParamSchema,
  generateHelper,
  inferRoleFromParts
};
