// @genkit-ai/ai/lib/formats/array.mjs
import "../chunk-7LUMLJGI.mjs";
import { GenkitError } from "@genkit-ai/core";
import { extractItems } from "../extract";
const arrayFormatter = {
  name: "array",
  config: {
    contentType: "application/json",
    constrained: true
  },
  handler: (request) => {
    var _a, _b, _c;
    if (((_a = request.output) == null ? void 0 : _a.schema) && ((_b = request.output) == null ? void 0 : _b.schema.type) !== "array") {
      throw new GenkitError({
        status: "INVALID_ARGUMENT",
        message: `Must supply an 'array' schema type when using the 'items' parser format.`
      });
    }
    let instructions;
    if ((_c = request.output) == null ? void 0 : _c.schema) {
      instructions = `Output should be a JSON array conforming to the following schema:
    
    \`\`\`
    ${JSON.stringify(request.output.schema)}
    \`\`\`
    `;
    }
    return {
      parseChunk: (chunk) => {
        var _a2;
        const cursor = ((_a2 = chunk.previousChunks) == null ? void 0 : _a2.length) ? extractItems(chunk.previousText).cursor : 0;
        const { items } = extractItems(chunk.accumulatedText, cursor);
        return items;
      },
      parseResponse: (response) => {
        const { items } = extractItems(response.text, 0);
        return items;
      },
      instructions
    };
  }
};
export {
  arrayFormatter
};
