// @genkit-ai/ai/lib/formats/jsonl.mjs
import "../chunk-7LUMLJGI.mjs";
import { GenkitError } from "@genkit-ai/core";
import JSON5 from "json5";
import { extractJson } from "../extract";
function objectLines(text) {
  return text.split("\n").map((line) => line.trim()).filter((line) => line.startsWith("{"));
}
const jsonlFormatter = {
  name: "jsonl",
  config: {
    contentType: "application/jsonl"
  },
  handler: (request) => {
    var _a, _b, _c, _d, _e, _f;
    if (((_a = request.output) == null ? void 0 : _a.schema) && (((_b = request.output) == null ? void 0 : _b.schema.type) !== "array" || ((_d = (_c = request.output) == null ? void 0 : _c.schema.items) == null ? void 0 : _d.type) !== "object")) {
      throw new GenkitError({
        status: "INVALID_ARGUMENT",
        message: `Must supply an 'array' schema type containing 'object' items when using the 'jsonl' parser format.`
      });
    }
    let instructions;
    if ((_f = (_e = request.output) == null ? void 0 : _e.schema) == null ? void 0 : _f.items) {
      instructions = `Output should be JSONL format, a sequence of JSON objects (one per line). Each line should conform to the following schema:

\`\`\`
${JSON.stringify(request.output.schema.items)}
\`\`\`
    `;
    }
    return {
      parseChunk: (chunk) => {
        var _a2;
        const results = [];
        const text = chunk.accumulatedText;
        let startIndex = 0;
        if ((_a2 = chunk.previousChunks) == null ? void 0 : _a2.length) {
          const lastNewline = chunk.previousText.lastIndexOf("\n");
          if (lastNewline !== -1) {
            startIndex = lastNewline + 1;
          }
        }
        const lines = text.slice(startIndex).split("\n");
        for (const line of lines) {
          const trimmed = line.trim();
          if (trimmed.startsWith("{")) {
            try {
              const result = JSON5.parse(trimmed);
              if (result) {
                results.push(result);
              }
            } catch (e) {
              break;
            }
          }
        }
        return results;
      },
      parseResponse: (response) => {
        const items = objectLines(response.text).map((l) => extractJson(l)).filter((l) => !!l);
        return items;
      },
      instructions
    };
  }
};
export {
  jsonlFormatter
};
