// genkit/lib/chat.mjs
import {
  __async,
  __spreadProps,
  __spreadValues
} from "./chunk-DFMI36TU.mjs";
import {
  runWithSession
} from "./session";
import { runInNewSpan } from "./tracing";
const MAIN_THREAD = "main";
class Chat {
  constructor(session, requestBase, options) {
    this.session = session;
    this.sessionId = options.id;
    this.threadName = options.thread;
    this.requestBase = requestBase == null ? void 0 : requestBase.then((rb) => {
      var _a, _b, _c;
      const requestBase2 = __spreadValues({}, rb);
      if (requestBase2 && requestBase2["prompt"]) {
        const basePrompt = requestBase2["prompt"];
        let promptMessage;
        if (typeof basePrompt === "string") {
          promptMessage = {
            role: "user",
            content: [{ text: basePrompt }]
          };
        } else if (Array.isArray(basePrompt)) {
          promptMessage = {
            role: "user",
            content: basePrompt
          };
        } else {
          promptMessage = {
            role: "user",
            content: [basePrompt]
          };
        }
        requestBase2.messages = [...(_a = requestBase2.messages) != null ? _a : [], promptMessage];
      }
      requestBase2.messages = [
        ...(_b = options.messages) != null ? _b : [],
        ...(_c = requestBase2.messages) != null ? _c : []
      ];
      this._messages = requestBase2.messages;
      return requestBase2;
    });
    this._messages = options.messages;
  }
  send(options) {
    return __async(this, null, function* () {
      return runWithSession(
        this.session,
        () => runInNewSpan({ metadata: { name: "send" } }, () => __async(this, null, function* () {
          var _a, _b;
          let resolvedOptions;
          let streamingCallback = void 0;
          if (typeof options === "string") {
            resolvedOptions = {
              prompt: options
            };
          } else if (Array.isArray(options)) {
            resolvedOptions = {
              prompt: options
            };
          } else {
            resolvedOptions = options;
            streamingCallback = resolvedOptions.streamingCallback;
          }
          let request = __spreadValues(__spreadProps(__spreadValues({}, yield this.requestBase), {
            messages: this.messages
          }), resolvedOptions);
          let response = yield this.genkit.generate(__spreadProps(__spreadValues({}, request), {
            streamingCallback
          }));
          this.requestBase = Promise.resolve(__spreadProps(__spreadValues({}, yield this.requestBase), {
            // these things may get changed by tools calling within generate.
            tools: (_a = response == null ? void 0 : response.request) == null ? void 0 : _a.tools,
            config: (_b = response == null ? void 0 : response.request) == null ? void 0 : _b.config
          }));
          yield this.updateMessages(response.messages);
          return response;
        }))
      );
    });
  }
  sendStream(options) {
    return runWithSession(
      this.session,
      () => runInNewSpan({ metadata: { name: "send" } }, () => __async(this, null, function* () {
        let resolvedOptions;
        if (typeof options === "string") {
          resolvedOptions = {
            prompt: options
          };
        } else if (Array.isArray(options)) {
          resolvedOptions = {
            prompt: options
          };
        } else {
          resolvedOptions = options;
        }
        const { response, stream } = yield this.genkit.generateStream(__spreadValues(__spreadProps(__spreadValues({}, yield this.requestBase), {
          messages: this.messages
        }), resolvedOptions));
        return {
          response: response.finally(() => __async(this, null, function* () {
            var _a, _b;
            const resolvedResponse = yield response;
            this.requestBase = Promise.resolve(__spreadProps(__spreadValues({}, yield this.requestBase), {
              // these things may get changed by tools calling within generate.
              tools: (_a = resolvedResponse == null ? void 0 : resolvedResponse.request) == null ? void 0 : _a.tools,
              config: (_b = resolvedResponse == null ? void 0 : resolvedResponse.request) == null ? void 0 : _b.config
            }));
            this.updateMessages(resolvedResponse.messages);
          })),
          stream
        };
      }))
    );
  }
  get genkit() {
    return this.session.genkit;
  }
  get messages() {
    var _a;
    return (_a = this._messages) != null ? _a : [];
  }
  updateMessages(messages) {
    return __async(this, null, function* () {
      this._messages = messages;
      yield this.session.updateMessages(this.threadName, messages);
    });
  }
}
export {
  Chat,
  MAIN_THREAD
};
