// @genkit-ai/tools-common/lib/cjs/types/eval.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatasetMetadataSchema = exports.DatasetSchemaSchema = exports.EvalRunSchema = exports.EvalKeyAugmentsSchema = exports.EvalRunKeySchema = exports.EvalResultSchema = exports.EvalMetricSchema = exports.EvalInputSchema = exports.DatasetSchema = exports.EvalInferenceInputSchema = exports.EvalInferenceStructuredInputSchema = void 0;
const zod_1 = require("zod");
exports.EvalInferenceStructuredInputSchema = zod_1.z.object({
    samples: zod_1.z.array(zod_1.z.object({
        testCaseId: zod_1.z.string().optional(),
        input: zod_1.z.any(),
        reference: zod_1.z.any().optional(),
    })),
});
exports.EvalInferenceInputSchema = zod_1.z.union([
    zod_1.z.array(zod_1.z.any()),
    exports.EvalInferenceStructuredInputSchema,
]);
exports.DatasetSchema = zod_1.z.array(zod_1.z.object({
    testCaseId: zod_1.z.string(),
    input: zod_1.z.any(),
    reference: zod_1.z.any().optional(),
}));
exports.EvalInputSchema = zod_1.z.object({
    testCaseId: zod_1.z.string(),
    input: zod_1.z.any(),
    output: zod_1.z.any(),
    error: zod_1.z.string().optional(),
    context: zod_1.z.array(zod_1.z.string()).optional(),
    reference: zod_1.z.any().optional(),
    traceIds: zod_1.z.array(zod_1.z.string()),
});
exports.EvalMetricSchema = zod_1.z.object({
    evaluator: zod_1.z.string(),
    score: zod_1.z.union([zod_1.z.number(), zod_1.z.string(), zod_1.z.boolean()]).optional(),
    rationale: zod_1.z.string().optional(),
    error: zod_1.z.string().optional(),
    traceId: zod_1.z.string().optional(),
    spanId: zod_1.z.string().optional(),
});
exports.EvalResultSchema = exports.EvalInputSchema.extend({
    metrics: zod_1.z.array(exports.EvalMetricSchema).optional(),
});
exports.EvalRunKeySchema = zod_1.z.object({
    actionRef: zod_1.z.string().optional(),
    datasetId: zod_1.z.string().optional(),
    datasetVersion: zod_1.z.number().optional(),
    evalRunId: zod_1.z.string(),
    createdAt: zod_1.z.string(),
});
exports.EvalKeyAugmentsSchema = exports.EvalRunKeySchema.pick({
    datasetId: true,
    datasetVersion: true,
    actionRef: true,
});
exports.EvalRunSchema = zod_1.z.object({
    key: exports.EvalRunKeySchema,
    results: zod_1.z.array(exports.EvalResultSchema),
    metricsMetadata: zod_1.z
        .record(zod_1.z.string(), zod_1.z.object({
        displayName: zod_1.z.string(),
        definition: zod_1.z.string(),
    }))
        .optional(),
});
exports.DatasetSchemaSchema = zod_1.z.object({
    inputSchema: zod_1.z
        .record(zod_1.z.any())
        .describe('Valid JSON Schema for the `input` field of dataset entry.')
        .optional(),
    referenceSchema: zod_1.z
        .record(zod_1.z.any())
        .describe('Valid JSON Schema for the `reference` field of dataset entry.')
        .optional(),
});
exports.DatasetMetadataSchema = zod_1.z.object({
    datasetId: zod_1.z.string(),
    size: zod_1.z.number(),
    schema: exports.DatasetSchemaSchema.optional(),
    targetAction: zod_1.z.string().optional(),
    version: zod_1.z.number(),
    createTime: zod_1.z.string(),
    updateTime: zod_1.z.string(),
});
