// @genkit-ai/ai/lib/message.mjs
import {
  __spreadProps,
  __spreadValues
} from "./chunk-7LUMLJGI.mjs";
import { extractJson } from "./extract";
class Message {
  static parseData(lenientMessage, defaultRole = "user") {
    if (typeof lenientMessage === "string") {
      return { role: defaultRole, content: [{ text: lenientMessage }] };
    }
    return __spreadProps(__spreadValues({}, lenientMessage), {
      content: Message.parseContent(lenientMessage.content)
    });
  }
  static parse(lenientMessage) {
    return new Message(Message.parseData(lenientMessage));
  }
  static parseContent(lenientPart) {
    if (typeof lenientPart === "string") {
      return [{ text: lenientPart }];
    } else if (Array.isArray(lenientPart)) {
      return lenientPart.map((p) => typeof p === "string" ? { text: p } : p);
    } else {
      return [lenientPart];
    }
  }
  constructor(message) {
    this.role = message.role;
    this.content = message.content;
    this.metadata = message.metadata;
  }
  /**
   * If a message contains a `data` part, it is returned. Otherwise, the `output()`
   * method extracts the first valid JSON object or array from the text contained in
   * the message and returns it.
   *
   * @returns The structured output contained in the message.
   */
  get output() {
    return this.data || extractJson(this.text);
  }
  toolResponseParts() {
    const res = this.content.filter((part) => !!part.toolResponse);
    return res;
  }
  /**
   * Concatenates all `text` parts present in the message with no delimiter.
   * @returns A string of all concatenated text parts.
   */
  get text() {
    return this.content.map((part) => part.text || "").join("");
  }
  /**
   * Returns the first media part detected in the message. Useful for extracting
   * (for example) an image from a generation expected to create one.
   * @returns The first detected `media` part in the message.
   */
  get media() {
    var _a;
    return ((_a = this.content.find((part) => part.media)) == null ? void 0 : _a.media) || null;
  }
  /**
   * Returns the first detected `data` part of a message.
   * @returns The first `data` part detected in the message (if any).
   */
  get data() {
    var _a;
    return (_a = this.content.find((part) => part.data)) == null ? void 0 : _a.data;
  }
  /**
   * Returns all tool request found in this message.
   * @returns Array of all tool request found in this message.
   */
  get toolRequests() {
    return this.content.filter(
      (part) => !!part.toolRequest
    );
  }
  /**
   * Converts the Message to a plain JS object.
   * @returns Plain JS object representing the data contained in the message.
   */
  toJSON() {
    let out = {
      role: this.role,
      content: [...this.content]
    };
    if (this.metadata) out.metadata = this.metadata;
    return out;
  }
}
export {
  Message
};
